### ISSUE TITLE
new T() where T : struct calls Activator.CreateInstance<T> which allocates on the heap

### ISSUE THREAD

**Issue by xoofx** [dotnet/coreclr#16621]
Even if we have a `struct` constraint, [`Activator.CreateInstance<T>`](https://github.com/dotnet/coreclr/blob/2825e34ed4b63482d1a2a1b1aeb6c341a4d05a44/src/mscorlib/src/System/Activator.cs#L75)  is going to call [`object RuntimeType.CreateInstanceDefaultCtor`](https://github.com/dotnet/coreclr/blob/08c87c57fb4fbff68abe4feb791417893c34dd2f/src/mscorlib/src/System/RtType.cs#L4802) which is causing `T` to be boxed to `object` before being returned.

Could we perform just at the beginning of `Activator.CreateInstance<T>` something like the following?

```C#
public static T CreateInstance<T>()
{
        if (typeof(T).IsValueType) return default(T);
```

**Comment by omariom** [comment]
`typeof(T).IsValueType` is not a JIT constant. 
But `default(T) != null` is. That's a bit ironic :)

**Comment by omariom** [comment]
Look like it is fixed [in corert](https://github.com/dotnet/corert/pull/3821)

**Comment by MichalStrehovsky** [comment]
```csharp
public static T CreateInstance<T>()
{
        if (typeof(T).IsValueType) return default(T);
```

Implementing it this way would regress #6843. This still needs to call the default constructor if there's any. CoreRT doesn't support structs with default constructors (because no mainstream .NET language has them and we currently focus on mainstream scenarios only there). CoreCLR does support them and C# would like to add them (the issue I'm referencing actually blocked the C# feature from shipping in the past).

I'm not saying it's not doable - it's just a bit more infrastructure to support this than an early out for valuetypes.

**Comment by roji** [comment]
See #7592 which would implement the intrinsics approach in coreclr - if done this issue would be obviated.

**Comment by YohDeadfall** [comment]
Resolved via #25145 by @jkotas.
